### ISSUE TITLE
[use_rethrow_when_possible] False positive in a switch expression

### ISSUE THREAD

**Issue by Levi-Lesches** [dart-lang/linter#4919]
**Describe the issue**

Rethrowing is supported in switch _statements_, [but not switch _expressions_](https://github.com/dart-lang/language/issues/3097). 

However, the lint still suggests to replace `throw error` with `rethrow`

**To Reproduce**

```dart
int getValue() => throw "Error2";

void main() {
  try {
    getValue();
  } catch (error) {
    final message = switch(error) {
      "Error1" => "Error 1",
      "Error2" => "Error 2",
      _ => throw error,
    };
    print(message);
  }
}
```

**Expected behavior**
The lint should not appear. 

**Additional context**
Given that `throw error` is allowed, I'm still not sure why `rethrow` shouldn't be. cc @munificent from the linked issue.

**Comment by munificent** [comment]
> Given that `throw error` is allowed, I'm still not sure why `rethrow` shouldn't be.

It's a language limitation. The bodies of switch expressions are expressions. Many years ago, we made `throw` an expression (mainly so you could use it as the body of `=>`-bodied methods), but we never turned `rethrow` into an expression. We probably should: https://github.com/dart-lang/language/issues/3097

**Comment by lrhn** [comment]
I think the lint _should_ appear in this case.
It (currently) requires you to rewrite the switch, but you probably should do that, so that the error is rethrown with the correct stack trace.


But we should make `rethrow` an expression. And `return`, `break` and `continue` too, because we can.

**Comment by srawlins** [comment]
Fixed by https://github.com/dart-lang/sdk/commit/51088718e194b6f8f97ff0faf41fa9f8a0dea28a
